File: src/main/java/com/jackmeng/stl/stl_Logger.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
// import java.lang.ref.Reference;
// import java.lang.ref.SoftReference;
import java.nio.file.Files;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.TimerTask;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * @author Jack Meng
 */
public class stl_Logger
    implements
        Runnable,
        Serializable
{
    public enum Logger_DisableBehavior
    {
        DROP_ALL, KEEP_ALL, CACHE_OLD;
    }

    private final long try_save_ms_periodic;
    private final String saveLocation, name;
    private final File file;
    private final Stack<String> logs = new Stack<>();
    private boolean running = false, enabled = true;
    private final AtomicBoolean saving = new AtomicBoolean(false);
    private final transient List<Runnable> afterRoutine;
    private transient long COLLECTED = 0L;
    private final AtomicLong CYCLE = new AtomicLong(0L);
    private int log_char_per_line = 70;
    //private final Reference<Stack<String>> logs_cache = new SoftReference<>(new Stack<>());
    //private Logger_DisableBehavior disableBehavior = Logger_DisableBehavior.KEEP_ALL;

    public stl_Logger(String loggerName, String saveLocationFolder, long save_time)
    {
        this.try_save_ms_periodic = save_time;
        this.name = loggerName;
        this.saveLocation = saveLocationFolder;
        this.file = new File(saveLocationFolder + stl0.dirm() + loggerName + "_" + stl_Chrono.format_millis() + ".log");
        afterRoutine = new ArrayList<>(10);
    }

    public stl_Logger(String loggerName, long save_time)
    {
        this(loggerName, stl_Commons.tmpdir(), save_time);
    }

    // public Logger_DisableBehavior disableBehavior()
    // {
    //     return disableBehavior;
    // }

    // public synchronized Logger_DisableBehavior disableBehavior(Logger_DisableBehavior e)
    // {
    //     Logger_DisableBehavior old = this.disableBehavior;
    //     this.disableBehavior = e;
    //     return old;
    // }

    public void enable(boolean e)
    {
        this.enabled = e;
    }

    public boolean enabled()
    {
        return this.enabled;
    }

    public void char_per_line(int i)
    {
        this.log_char_per_line = i;
    }

    public int char_per_line()
    {
        return log_char_per_line;
    }

    public stl_Logger(String loggerName, String saveLocation)
    {
        this(loggerName, saveLocation, 4000L);
    }

    public synchronized void push(Object contents)
    {
        String placeholder = String.format("%09d", COLLECTED + 1) + " | " + String.format("%09d", CYCLE.get()) + "   |   " + stl_Chrono.format_millis("HH:mm:ss MM/dd/YYYY") + "    |    ";
        if(saving.get())
            afterRoutine.add(() -> logs.push(placeholder + stl_Str.insert_nl(contents.toString(), char_per_line(), "\n" + stl_Str.n_copies(placeholder.length() - 5, " ")  + "|    ")));
        else
            logs.push(placeholder + stl_Str.insert_nl(contents.toString(), char_per_line(), "\n" + stl_Str.n_copies(placeholder.length() - 5, " ") + "|    "));
        COLLECTED++;
    }

    public synchronized void log(Object ... contents)
    {
        for(Object e : contents) push(e);
    }

    public synchronized void kill()
    {
        afterRoutine.clear();
        logs.clear();
        stl0.STL_TIMER0.cancel();
        Thread.currentThread().interrupt();
    }

    public String name()
    {
        return name;
    }

    public String saveLocation()
    {
        return saveLocation;
    }

    @Override
    public synchronized void run()
    {
        if(!running && enabled)
        {
            if(file.exists())
                file.delete();
            try
            {
                file.createNewFile();
            }
            catch (IOException e)
            {
                e.printStackTrace();
            }

            stl0.STL_TIMER0.schedule(new TimerTask()
            {
                @Override
                public void run()
                {
                    if(enabled)
                    {
                        if(logs.size() > 0)
                        {
                            saving.set(true);
                            synchronized (logs) {
                                StringBuilder sb = new StringBuilder();
                                while(!logs.isEmpty())
                                    sb.append(logs.pop()).append('\n');

                                try {
                                    Files.write(file.toPath(), sb.toString().getBytes(), StandardOpenOption.WRITE, StandardOpenOption.APPEND);
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }
                                if (afterRoutine.size() > 0)
                                    for(int i = 0; i < afterRoutine.size(); i++)
                                        afterRoutine.remove(i).run();
                            }
                            saving.set(false);
                            CYCLE.set(CYCLE.get() + 1L);
                        }
                    }
                }
            }, 500L, try_save_ms_periodic);
            running = true;
        }
    }
}

File: src/main/java/com/jackmeng/stl/stl_Logger.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import com.jackmeng.stl.stl0.Companion.dirm
import com.jackmeng.stl.stl_Chrono.format_millis
import com.jackmeng.stl.stl_Commons.tmpdir
import java.lang.Runnable
import java.io.File
import java.util.Stack
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import kotlin.jvm.Synchronized
import java.io.IOException
import java.io.Serializable
import java.util.TimerTask
import java.lang.StringBuilder
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import java.util.ArrayList

// import java.lang.ref.Reference;
// import java.lang.ref.SoftReference;
/**
 * @author Jack Meng
 */
class stl_Logger(private val name:String , private val saveLocation:String , private val try_save_ms_periodic:Long):
		Runnable , Serializable
{
	enum class Logger_DisableBehavior
	{
		DROP_ALL , KEEP_ALL , CACHE_OLD
	}
	
	private val file:File=File(saveLocation+dirm()+name+"_"+format_millis()+".log")
	private val logs=Stack<String>()
	private var running=false
	private var enabled=true
	private val saving=AtomicBoolean(false)
	
	@Transient
	private val afterRoutine:MutableList<Runnable>
	
	@Transient
	private var COLLECTED=0L
	private val CYCLE=AtomicLong(0L)
	private var log_char_per_line=70
	
	//private final Reference<Stack<String>> logs_cache = new SoftReference<>(new Stack<>());
	//private Logger_DisableBehavior disableBehavior = Logger_DisableBehavior.KEEP_ALL;
	init
	{
		afterRoutine=ArrayList(10)
	}
	
	constructor(loggerName:String , save_time:Long):this(loggerName , tmpdir() , save_time)
	
	// public Logger_DisableBehavior disableBehavior()
	// {
	//     return disableBehavior;
	// }
	// public synchronized Logger_DisableBehavior disableBehavior(Logger_DisableBehavior e)
	// {
	//     Logger_DisableBehavior old = this.disableBehavior;
	//     this.disableBehavior = e;
	//     return old;
	// }
	fun enable(e:Boolean)
	{
		enabled=e
	}
	
	fun enabled():Boolean
	{
		return enabled
	}
	
	fun char_per_line(i:Int)
	{
		log_char_per_line=i
	}
	
	fun char_per_line():Int
	{
		return log_char_per_line
	}
	
	constructor(loggerName:String , saveLocation:String):this(loggerName , saveLocation , 4000L)
	
	@Synchronized
	fun push(contents:Any)
	{
		val placeholder=String.format("%09d" , COLLECTED+1)+" | "+String.format(
			"%09d" ,
			CYCLE.get()
		)+"   |   "+format_millis("HH:mm:ss MM/dd/YYYY")+"    |    "
		if (saving.get()) afterRoutine.add(Runnable {
			logs.push(
				placeholder+stl_Str.insert_nl(
					contents.toString() , char_per_line() , """
 	
 	${stl_Str.n_copies(placeholder.length-5 , " ")}|    
 	""".trimIndent()
				)
			)
		})
		else logs.push(
			placeholder+stl_Str.insert_nl(
				contents.toString() , char_per_line() , """
 	
 	${stl_Str.n_copies(placeholder.length-5 , " ")}|    
 	""".trimIndent()
			)
		)
		COLLECTED++
	}
	
	@Synchronized
	fun log(vararg contents:Any)
	{
		for (e in contents) push(e)
	}
	
	@Synchronized
	fun kill()
	{
		afterRoutine.clear()
		logs.clear()
		stl0.STL_TIMER0.cancel()
		Thread.currentThread().interrupt()
	}
	
	fun name():String
	{
		return name
	}
	
	fun saveLocation():String
	{
		return saveLocation
	}
	
	@Synchronized
	override fun run()
	{
		if (!running&&enabled)
		{
			if (file.exists()) file.delete()
			try
			{
				file.createNewFile()
			} catch (e:IOException)
			{
				e.printStackTrace()
			}
			stl0.STL_TIMER0.schedule(object:TimerTask()
			{
				override fun run()
				{
					if (enabled)
					{
						if (logs.size>0)
						{
							saving.set(true)
							synchronized(logs) {
								val sb=StringBuilder()
								while (!logs.isEmpty()) sb.append(logs.pop()).append('\n')
								try
								{
									Files.write(
										file.toPath() ,
										sb.toString().toByteArray() ,
										StandardOpenOption.WRITE ,
										StandardOpenOption.APPEND
									)
								} catch (e:IOException)
								{
									e.printStackTrace()
								}
								if (afterRoutine.size>0) for (i in afterRoutine.indices) afterRoutine.removeAt(i).run()
							}
							saving.set(false)
							CYCLE.set(CYCLE.get()+1L)
						}
					}
				}
			} , 500L , try_save_ms_periodic)
			running=true
		}
	}
}